lib/ansible/modules/source_control/gitlab_project_variable.py

```diff
@@ -0,0 +1,246 @@
+from ansible.module_utils.api import basic_auth_argument_spec
+
+
+GITLAB_IMP_ERR = None
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except Exception:
+    GITLAB_IMP_ERR = traceback.format_exc()
+    HAS_GITLAB_PACKAGE = False
+
+
+class gitlab_project_variables(object):
+
+    def __init__(self, module):
+        self.repo = gitlab.Gitlab(module.params['api_url'],
+                                  private_token=module.params['api_token'])
+        self.project = self.get_project(module.params['project'])
+        self._module = module
+
+    def auth(self):
+        self.repo.auth()
+
+    def get_project(self, project_name):
+        return self.repo.projects.get(project_name)
+
+    def list_all_project_variables(self):
+        raw_variable_list = self.project.variables.list()
+        retval = []
+        if len(raw_variable_list) > 0:
+            for item in raw_variable_list:
+                retval.append(item.get_id())
+        return retval
+
+    def create_variable(self, key, value):
+        if self._module.check_mode:
+            return
+        return self.project.variables.create({"key": key, "value": value})
+
+    def update_variable(self, key, value):
+        var = self.project.variables.get(key)
+        if var.value == value:
+            return False
+        if self._module.check_mode:
+            return True
+        var.save()
+        return True
+
+    def delete_variable(self, key):
+        if self._module.check_mode:
+            return
+        return self.project.variables.delete(key)
+
+
+def native_python_main(this_gitlab, purge, var_list, state):
+
+    change = False
+    return_value = dict(added=list(), updated=list(), removed=list())
+
+    existing_variables = this_gitlab.list_all_project_variables()
+
+    for key in var_list:
+        if key in existing_variables and state == 'present':
+            change = this_gitlab.update_variable(
+                key, var_list[key]) or change
+            pop_index = existing_variables.index(key)
+            existing_variables.pop(pop_index)
+            return_value['updated'].append(key)
+        elif key not in existing_variables and state == 'present':
+            this_gitlab.create_variable(key, var_list[key])
+            change = True
+            return_value['added'].append(key)
+        elif key in existing_variables and state == 'absent':
+            this_gitlab.delete_variable(key)
+            change = True
+            return_value['removed'].append(key)
+
+    if len(existing_variables) > 0 and purge:
+        for item in existing_variables:
+            this_gitlab.delete_variable(item)
+            change = True
+            return_value['removed'].append(item)
+
+    existing_variables = this_gitlab.list_all_project_variables()
+    return change, return_value
+
+
+def main():
+    argument_spec = basic_auth_argument_spec()
+    argument_spec.update(
+        api_token=dict(required=True, no_log=True, type='str'),
```

Comment:
Simplify this by not adding required=False, and always starting with the type.
Since every parameter has a type, starting with this adds consistency.

Reply:
@Lunik but the api_token is required. Why should it be set to false?

Reply:
On all your parameters, if there are not required, there is no need to specify the `required=False`.
But if they are, you should speficy `required=True`

Reply:
`api_token` is a required parameter